### PR TITLE
Network::Connection: Add L4 crash dumping support

### PR DESCRIPTION
**Pull request by KBaichoo**
Added ScopeTrackedObject to Network::Connection so we can dump L4 information on crashes during network level processing. 

This will be followed up with PRs changing ScopeTrackedObject to a stack so that we can 'unwind' the various network objects on a fatal crash. 

Signed-off-by: Kevin Baichoo <kbaichoo@google.com>

Commit Message: Network::Connection: Add L4 crash dumping support.
Additional Description:
Risk Level: Medium
Testing: Unit Tests
Docs Changes: Included
Release Notes: Included
Platform Specific Features: NA
Related Issue: #14249

### CODE REVIEW

File: include/envoy/common/scope_tracker.h [hunk 1]
```diff
@@ -7,16 +7,19 @@
...
 /*
- * A class for tracking the scope of work.
- * Currently this is only used for best-effort tracking the L7 stream doing
- * work if a fatal error occurs.
+ * An interface for tracking the scope of work. Implementors of this interface
+ * can be registered to the dispatcher when they're active on the stack. If a
+ * fatal error occurs while they were active, the dumpState method will be
+ * called.
+ *
+ * Currently this is only used for the L4 network connection and L7 stream.
```
Comment (antoniovicente) [inline include/envoy/common/scope_tracker.h:15]: class ScopeTrackedObject : public Dumpable {}

But are the two interfaces really needed?  Why is it important to distinguish between objects that can be registered and ones that simply have a dump method?

File: test/common/network/connection_impl_test.cc [hunk 2]
```diff
@@ -1851,6 +1852,50 @@ TEST_P(ConnectionImplTest, DelayedCloseTimeoutNullStats) {
...
+        new Network::Address::Ipv6Instance("::1:2:3:4", 56789, nullptr)};
+  }
+
+  auto connection_socket =
+      std::make_unique<ConnectionSocketImpl>(std::move(io_handle), local_addr, server_addr);
+  connection_socket->setRequestedServerName("envoyproxy.io");
+
+  // Start measuring memory and dump state.
+  Stats::TestUtil::MemoryTest memory_test;
+  connection_socket->dumpState(std::cout, 0);
```
Comment (antoniovicente) [inline test/common/network/connection_impl_test.cc:1878]: Would be good to verify the contents of this dump.  Could you use the mechanism that allows dumping to a stack-allocated array?

File: source/common/network/connection_impl.cc [hunk 3]
```diff
@@ -753,6 +756,27 @@ void ConnectionImpl::flushWriteBuffer() {
...
+    return os << "Closed";
+  }
+  return os;
+}
+
+// ScopeTrackedObject
+void ConnectionImpl::dumpState(std::ostream& os, int indent_level) const {
+  const char* spaces = spacesForLevel(indent_level);
+  os << spaces << "ConnectionImpl " << this << DUMP_MEMBER(connecting_) << DUMP_MEMBER(bind_error_)
+     << ", state() :" << state() << DUMP_MEMBER(read_buffer_limit_) << "\n";
```
Comment (antoniovicente) [inline source/common/network/connection_impl.cc:775]: Would this work?

<< DUMP_MEMBER(state()) <<

File: source/common/common/dump_state_utils.h [hunk 4]
```diff
@@ -9,11 +9,26 @@ namespace Envoy {
...
+
+// This is a workaround for fact that MSVC expands __VA_ARGS__ after passing them into a macro,
+// rather than before passing them into a macro. Without this,
+// _DUMP_MEMBER_SELECTOR does not work correctly when compiled with MSVC.
+#define EXPAND(X) X
+
+// If DUMP_MEMBER is called with one argument, then _DUMP_MEMBER is called.
+// If DUMP_MEMBER is called with two arguments, then _DUMP_MEMBER_VIA_VALUE is called.
+#define DUMP_MEMBER(...)                                                                           \
+  EXPAND(_DUMP_MEMBER_SELECTOR(__VA_ARGS__, _DUMP_MEMBER_VIA_VALUE, _DUMP_MEMBER)(__VA_ARGS__))
```
Comment (antoniovicente) [inline source/common/common/dump_state_utils.h:24]: I'm not seeing any uses of the 2 argument DUMP_MEMBER macro.  I think this was done prior to the introduction of DUMP_NULLABLE_MEMBER below

**Review by antoniovicente — COMMENTED** [review]
Thanks for dumping some info about the connection on crash.  Looking forward to the followup change that allows multiple objects to be dumped on crash, so we can get both the L4 and L7 state on failure.

File: test/common/network/connection_impl_test.cc (hunk 2 above)
Comment (KBaichoo) [reply test/common/network/connection_impl_test.cc]: Done.

File: include/envoy/common/scope_tracker.h (hunk 1 above)
Comment (antoniovicente) [reply include/envoy/common/scope_tracker.h:15]: Is not being able to register the object useful?  Having 2 interfaces that are so similar adds to cognitive load.

File: source/common/network/connection_impl.cc [hunk 5]
```diff
@@ -753,6 +756,27 @@ void ConnectionImpl::flushWriteBuffer() {
...
+    return os << "Open";
+  case Connection::State::Closing:
+    return os << "Closing";
+  case Connection::State::Closed:
+    return os << "Closed";
+  }
+  return os;
+}
+
+// ScopeTrackedObject
```
Comment (antoniovicente) [inline source/common/network/connection_impl.cc:771]: Please remove comment. We don't seem to follow the practice of tagging interface methods in cc files.

File: source/common/network/connection_impl.cc [hunk 6]
```diff
@@ -753,6 +756,27 @@ void ConnectionImpl::flushWriteBuffer() {
   }
 }
 
+std::ostream& operator<<(std::ostream& os, Connection::State connection_state) {
```
Comment (antoniovicente) [inline source/common/network/connection_impl.cc:759]: Implementing this operator in the middle of this cc file is a bit odd.  Consider moving to an anonymous namespace at the top of the file or to the library that defines the Connection::State enum class.

**Review by antoniovicente — APPROVED** [review]
Changes look fine to me and are tested. Off to @envoyproxy/senior-maintainers for review.

File: include/envoy/common/scope_tracker.h (hunk 1 above)
Comment (mattklein123) [reply include/envoy/common/scope_tracker.h:15]: Yeah I would tend to agree that the two interfaces do not feel that useful to me, but I don't feel strongly about it. At minimum I would derive from the other class as @antoniovicente suggests.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Generally LGTM other than the small comments. Can you also merge main to get CI to pass?

/wait

File: source/common/network/connection_impl.cc (hunk 6 above)
Comment (KBaichoo) [reply source/common/network/connection_impl.cc]: Moved it to anonymous namespace at the top of this file. Done.

File: include/envoy/common/scope_tracker.h (hunk 1 above)
Comment (KBaichoo) [reply include/envoy/common/scope_tracker.h:15]: Fair point, will just go with `ScopeTrackedObject` to keep it simple

File: docs/root/version_history/current.rst [hunk 7]
```diff
@@ -68,6 +68,7 @@ New Features
...
 * config: added ability to flush stats when the admin's :ref:`/stats endpoint <operations_admin_interface_stats>` is hit instead of on a timer via :ref:`stats_flush_on_admin <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.stats_flush_on_admin>`.
 * config: added new runtime feature `envoy.features.enable_all_deprecated_features` that allows the use of all deprecated features.
+* crash support: added the ability to dump L4 data on crash.
```
Comment (antoniovicente) [inline docs/root/version_history/current.rst:71]: nit: change "L4 data" to "L4 connection data"

**Review by antoniovicente — DISMISSED** [review]
Thanks!

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**